### PR TITLE
feat(vllm): serve GridFM PowerFlow reconstruction via vLLM /pooling

### PR DESCRIPTION
**Pull request by romeokienzler**
## Serve GridFM PowerFlow reconstruction via vLLM `/pooling`

Adds a `gridfm_graphkit.vllm` subpackage that serves a trained `GNS_heterogeneous`
model through vLLM's pooling endpoint as an attention-free multimodal model,
mirroring vLLM's Terratorch wrapper.

### What's in it
- **`model.py`** — `GridFMForPooling` wraps the GNN. The graph arrives as a bag of
  named tensors under one multimodal modality; `forward` loops over the collation's
  leading item dimension (1 graph per prompt token; warmup replicates the dummy
  across `max_num_reqs`), runs the GNN with `return_embeddings=True`, and returns
  one packed row per item so the leading dim equals the token count vLLM
  slices/indexes against.
- **`graph_codec.py`** — single source of truth flattening `HeteroData` ↔ named
  tensors and packing predictions + embeddings into one opaque tensor for the
  identity pooler.
- **`config.py` / `export.py`** — rebuild model + normalizer + transforms from a
  `pretrained_cfg` payload and write an HF/vLLM-loadable model directory
  (`model_type`-less config → `TimmWrapperConfig`, routed to `GridFMGNS`).
- **`plugins/`** — `vllm.general_plugins` registers the `GridFMGNS` architecture;
  `vllm.io_processor_plugins` registers the PowerFlow-reconstruction IO processor.
- Extracts `build_hetero_data` into `datasets/graph_builder.py` so the IO processor
  reuses the exact graph-building path as training.

Adds a `[vllm]` optional extra (`vllm>=0.26,<0.27`, `safetensors`, `timm`) and the
two entry-point tables. Version bumped to **0.9.1**.

### Testing
Offline unit/integration tests cover the codec round-trip and the full serving core
(build → normalize → mask → encode → decode → forward → pack → unpack → denormalize).

Verified end-to-end on an **H100** (bluevela): a case14 case POSTed to `/pooling`
returns **HTTP 200** with per-node embeddings and Vm/Va/P/Q + Pg predictions —
`bus_predictions 14×4`, `gen_predictions 5×1`, `bus_embeddings 14×32`,
`gen_embeddings 5×32`. Full CPU test suite (6 tests) + GPU serve e2e both green
against this branch's tip.

### Note
Targets `vllm>=0.26,<0.27`. Downstream (algorithm-nexus) wires `gridfm-graphkit[vllm]`
into ecosystem + candidate targets once 0.9.1 is published to PyPI.

---
cc @albanpuech for review 🙏

### CODE REVIEW

**Comment by romeokienzler** [comment]
@romeokienzler thanks for this — the vLLM serving subpackage is nicely structured, deps are declared under a `[vllm]` optional extra, and it ships codec round-trip + pipeline tests. A few things before a maintainer takes a look:

**What's needed**
- **DCO is failing** — neither commit carries a `Signed-off-by` line. Please sign off all commits (e.g. `git commit -s --amend` then `git rebase --signoff` / force-push) so the DCO check goes green.
- **Version bump to 0.9.1** — the second commit bumps `pyproject.toml` from 0.9.0 → 0.9.1. Releases are usually cut separately by a maintainer; consider dropping the bump from this feature PR (or flag it explicitly so it's an intentional call).
- **Pending CI** — `pytests`, `pre-commit-run`, `CodeQL (Python)` and the `security`/`pip-audit` jobs are still running; worth confirming they land green. Note: if `pip-audit` trips on the known lightning advisory (PYSEC-2026-3624), that's a repo-wide infra blocker being fixed separately — **not on you**.
- **Scope check** — extending the core with a serving path that reuses the exact training graph-builder and stays behind an optional extra reads as reasonable core functionality, but per CONTRIBUTING's "Scope" section I'll defer to a maintainer to confirm this belongs here vs. a companion applications repo.

Nothing blocking beyond the DCO signoff. Looks close.

 — 🤖 _automated pre-review; a maintainer will follow up_

File: gridfm_graphkit/datasets/powergrid_hetero_dataset.py [hunk 1]
```diff
@@ -317,116 +317,7 @@ def _build_and_save_scenario(
...
-            gen_df[["gen_index", "bus"]].values.T.copy(),
-            dtype=torch.long,
-        )
-        data["bus", "connected_to", "gen"].edge_index = torch.tensor(
-            gen_df[["bus", "gen_index"]].values.T.copy(),
-            dtype=torch.long,
-        )
-
-        data["scenario_id"] = torch.tensor([scenario], dtype=torch.long)
+        data = build_hetero_data(bus_df, gen_df, branch_df, scenario)
```
Comment (romeokienzler) [inline gridfm_graphkit/datasets/powergrid_hetero_dataset.py:320]: @albanpuech has this been merged correctly?

Comment (albanpuech) [reply gridfm_graphkit/datasets/powergrid_hetero_dataset.py:320]: sorry, wdym?

Comment (romeokienzler) [reply gridfm_graphkit/datasets/powergrid_hetero_dataset.py:320]: Sorry, that was too terse from me — my worry was about the merge of `main` into this branch, not about your code.

This PR moves the whole `HeteroData` construction out of `_build_and_save_scenario` into a new `gridfm_graphkit/datasets/graph_builder.py` (`build_hetero_data`), so the serving path and training build graphs from one source. Meanwhile your #117 (`fix(datasets): copy transposed edge arrays to avoid negative strides`) patched that same code in its *old* location on `main`. When a fix and a file move meet in a merge, git can resolve it cleanly and still silently drop the fix — that is what I wanted to double-check.

It came through correctly. All four `.copy()` calls are in the new home:

```
graph_builder.py:108    branch_df[["from_bus", "to_bus"]].values.T.copy(),
graph_builder.py:116    branch_df[["to_bus", "from_bus"]].values.T.copy(),
graph_builder.py:137    gen_df[["gen_index", "bus"]].values.T.copy(),
graph_builder.py:141    gen_df[["bus", "gen_index"]].values.T.copy(),
```

and no `.values.T` is left behind in `powergrid_hetero_dataset.py`. I also diffed the moved block against `main`'s version line by line — it is equivalent apart from four intentional changes:

- feature lists became module-level constants (`BUS_FEATURES`, `GEN_FEATURES`, `FORWARD_BRANCH_FEATURES`, ...) so the serving codec can import them;
- `build_hetero_data` returns the graph and `_build_and_save_scenario` keeps the `torch.save`;
- `gen_df.reset_index(drop=True)` instead of `reset_index()` — avoids adding an unused `index` column; `gen_index` is still `0..N-1`, so the edge indices are identical;
- `torch.arange(len(bus_df)).numpy()` in the bus-ordering assert, to keep that a plain numpy comparison.

#117 was the only commit from `main` touching this file; the only other file both sides changed was `pyproject.toml` (version, now `0.9.0.2`), and `pytests` is green on the merged branch.

Resolving this — thanks for bearing with the cryptic question.

**Comment by romeokienzler** [comment]
@albanpuech when you have a moment, would you mind taking a look at this one? No rush if you are deep in something else — just do not want it to go stale.

Where it stands:

- The `powergrid_hetero_dataset.py` thread is answered and resolved: your #117 `.copy()` fix survived the move into `graph_builder.py`, and the moved block is otherwise equivalent to `main` (details in the thread).
- The branch is level with `main` and all CI is green except DCO, which needs a sign-off rebase on the two early commits — I will push that separately, it does not touch the code you would be reviewing.
- Version now takes `0.9.0.2` rather than `0.9.1`, since `main` moved to `0.9.0.1` in #121 and this PR only adds an optional extra.

The parts most worth your eyes are `gridfm_graphkit/datasets/graph_builder.py` (the extraction from the dataset class) and `gridfm_graphkit/vllm/graph_codec.py` (the serving-side encode/decode that has to match it). Everything else sits behind the optional `[vllm]` extra and is inert without vLLM installed.

Thanks!

**Review by paolofraccaro — APPROVED** [review]

**Comment by romeokienzler** [comment]
## Vela integration tests pass on this branch

Ran `integrationtests` on Vela for `feat/vllm-serving` — job **1741972**, single H100 (`p4-r15-n4`), queue `normal`, **2 passed in 36:40**, LSF `Done successfully`.

All 11 metrics inside the calibrated bounds on **attempt 1**:

| metric | value | bounds |
|---|---|---|
| PF — PBE Mean | 0.4140397 | [0.4098993, 0.4181801] |
| Avg. active res. (MW) | 0.2915160 | [0.2886008, 0.2944312] |
| Avg. reactive res. (MVar) | 0.1052882 | [0.1042353, 0.1063411] |
| Mean Qg violation | 0.1033714 | [0.1023377, 0.1044051] |
| Mean Qg violation PV buses | 0.0816202 | [0.0808040, 0.0824364] |
| Mean Qg violation REF buses | 0.1865719 | [0.1847062, 0.1884376] |
| Mean branch angle difference violation (rad) | 0.0 | [0.0, 0.0] |
| Mean branch thermal violation from (MVA) | 0.0 | [0.0, 0.0] |
| Mean branch thermal violation to (MVA) | 0.0 | [0.0, 0.0] |
| Mean optimality gap (%) | 1.5786682 | [1.5628816, 1.5944549] |
| RMSE PG generators (MW) | 2.6503038 | [2.6238008, 2.6768069] |

### Why this run is the interesting one for this PR

The `data_out` staging had no `processed/` cache, so all ~20k scenario graphs were rebuilt from `raw/` through the extracted `graph_builder.build_hetero_data` rather than loaded from `.pt` files written by the pre-refactor code. The metrics still land inside bounds that were calibrated **before** the extraction, which is numerical evidence that moving the graph construction out of `HeteroGridDatasetDisk` is behaviour-preserving — the concern raised in the `powergrid_hetero_dataset.py` thread.

The vLLM code itself is not exercised by these two tests (they only run `test_train_pf` / `test_train_opf`); it stays behind the optional `[vllm]` extra and is covered by `tests/test_vllm_graph_codec.py` and `tests/test_vllm_pipeline.py` in the normal CI run.

### Expected warning

The usual fingerprint mismatch fired — bounds were calibrated on a different environment, so it warns but still asserts:

```
gridfm_datakit_version:   baseline='1.0.5rc1'  current='1.1.0'
gridfm_graphkit_version:  baseline='0.8.0'     current='0.9.0.2'
hostname:                 baseline='p4-r19-n2' current='p4-r15-n4'
```

`integrationtests/calibration_baseline.json` was left untouched (the on-machine copy is excluded from the sync), so these bounds remain the ones from the 0.8.0 calibration.

### DCO

Also pushed the sign-off I promised: the two early commits (`feat(vllm): ...` and the version bump) now carry `Signed-off-by`, added via a message-only rewrite so the trees are byte-identical — the PR diff is unchanged at 17 files, +1595/-113. DCO is green.
